cmd/exit_handler.go

```diff
@@ -0,0 +1,312 @@
+	for _, signal := range s.signalHandlers {
+		result = append(result, signal.GetSignal())
+	}
+	return result
+}
+
+// AddDeferFunc appends new defer function (with priority) for execution
+func (s *ExitHandler) AddDeferFunc(input DeferFunction) {
+	inputHasLastPriority := input.GetPriority() == Last
+	for _, deferFunc := range s.deferFunctions {
+		if deferFunc.GetPriority() == Last && inputHasLastPriority {
+			panic("defer function with 'Last' priority has been already specified")
+		}
+	}
+	s.deferFunctions = append(s.deferFunctions, input)
+}
+
+// ExitZero is a single point for exiting from the service with 0 code
+func (s *ExitHandler) ExitZero() {
+	s.gracefulExit()
+	os.Exit(0)
+}
+
+// ExitOne is a single point for exiting from the service with 1 code
+func (s *ExitHandler) ExitOne() {
+	s.gracefulExit()
+	os.Exit(1)
+}
+
+func (s *ExitHandler) gracefulExit() {
+	// at first we finalize our handlers (close listeners and call callbacks)
+	s.executeFinalizeOnHandlers()
+	// finally we call defer functions
+	s.executeDeferFunctions()
+}
+
+func (s *ExitHandler) executeDeferFunctions() {
+	var lastDefer DeferFunction
+	for _, deferFunction := range s.deferFunctions {
+		if deferFunction.GetPriority() == Last {
+			lastDefer = deferFunction
+		}
+		deferFunction.Call()
+	}
+	// defer function with last priority has not been found
+	if lastDefer == nil {
+		return
+	}
+	lastDefer.Call()
+}
+
+func (s *ExitHandler) executeFinalizeOnHandlers() {
+	for _, handler := range s.signalHandlers {
+		handler.Finalize()
+	}
+}
+
+// AcraSignal is a common interface for system signal handlers in Acra
+type AcraSignal interface {
+	Finalize()
+	GetSignal() os.Signal
+	AddCallback(callback *SignalCallback_)
+	AddListener(listener net.Listener)
+}
+
+// SigTERMHandler is an implementation of AcraSignal for SIGTERM signal
+type SigTERMHandler struct {
+	listeners []net.Listener
+	callbacks []*SignalCallback_
+}
+
+// AddCallback appends callback with priority to SigTERMHandler
+func (s *SigTERMHandler) AddCallback(input *SignalCallback_) {
+	inputHasLastPriority := input.GetPriority() == Last
+	for _, callback := range s.callbacks {
+		if callback.GetPriority() == Last && inputHasLastPriority {
+			panic("callback with 'Last' priority has been already specified")
+		}
+	}
+	s.callbacks = append(s.callbacks, input)
+}
+
+// AddListener appends listener to SigTERMHandler
+func (s *SigTERMHandler) AddListener(listener net.Listener) {
+	s.listeners = append(s.listeners, listener)
+}
+
+// Finalize closes all listeners and executes callbacks with priority considering
+func (s *SigTERMHandler) Finalize() {
+	for _, listener := range s.listeners {
+		listener.Close()
+	}
+	var lastCallback *SignalCallback_
+	for _, callback := range s.callbacks {
+		if callback.GetPriority() == Last {
+			lastCallback = callback
+		}
+		callback.Call()
+	}
+	// defer function with last priority has not been found
+	if lastCallback == nil {
+		return
+	}
+	lastCallback.Call()
+}
```

Comment:
This seems to be awfully similar to `func (s *ExitHandler) executeDeferFunctions()`.